codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/endpoint/EndpointRulesetIndex.kt

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rust.codegen.client.smithy.endpoint
+
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.knowledge.KnowledgeIndex
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.rulesengine.language.EndpointRuleSet
+import software.amazon.smithy.rulesengine.traits.EndpointRuleSetTrait
+import software.amazon.smithy.rust.codegen.core.util.getTrait
+
+class EndpointRulesetIndex(model: Model) : KnowledgeIndex {
+
+    private val rulesets: HashMap<ServiceShape, EndpointRuleSet?> = HashMap()
```

Comment:
Since codegen is parallelized by Smithy, does this need to be `ConcurrentHashMap`?